InstructorFeedbackResultsPageScalabilityTest: tests data should not be removed and refreshed for each test case #7171

Malkone opened the pull request:
Fixes #7171

**PR Checklist**

To reduce number of calls to the datastore when performing scalability test following changes are introduced:
 - instead of 6 json files with different number of students and questions 1 json file is created with maximum number of students and questions (in class InstructorFeedbackResultsPageDataGenerator)
 - in class InstructorFeedbackResultsPageScalabilityTest instead of importing separate json file for each test in ```refreshTestData()``` method, now one json file is imported in additional method ```loadTestData()```, where also non-modified test data is initiated.  For this purpose additional DataBundle is created which contains test data for maximum number of students and questions
 - in method ```refreshTestData()``` DataBundle is updated by adding extra data or removing excessive data. This method uses new additional methods ```increaseNumOfStudents()```, ```increaseNumOfQuestions()```, ```decreaseNumOfStudents()```, ```decreaseNumOfQuestions()``` and ```updateFeedbackResponses()```

Malkone commented:
@samsontmr sure, I can change ```if(condition)``` to ```if(!condition)``` and place the code inside. Actually, I read that using continue is not very good and I did it with ```continue``` just to follow rules, described in [this article](https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88), because during my previous pull-request I was advised by @damithc to follow its rules and "keep the 'happy path' unindented".

Malkone commented:
@samsontmr I removed now all uses of continue,

samsontmr commented on src/client/java/teammates/client/scripts/scalabilitytests/InstructorFeedbackResultsPageDataGenerator.java:
```diff
@@ -191,21 +192,17 @@ private InstructorFeedbackResultsPageDataGenerator(int numQuestions, int numStud
-        //Number of students and questions for each data set.
-        int[] studentNums = {10, 20};
-        int[] questionNums = {1, 5, 10};
+        // One set of maximum number of students and questions to create single data set for all tests
```
`Maximum number of students and questions to create a single data set for all tests.`

samsontmr commented on src/client/java/teammates/client/scripts/scalabilitytests/InstructorFeedbackResultsPageScalabilityTest.java:
```diff
@@ -18,28 +24,153 @@
+        }
+    }
+
+    private void decreaseNumOfStudents(int remainingNumOfStudents) {
```
Consider renaming this method: its similarity to `increaseNumOfStudents` suggests it decreases the number of students by the integer passed in.

samsontmr commented on src/client/java/teammates/client/scripts/scalabilitytests/InstructorFeedbackResultsPageScalabilityTest.java:
```diff
@@ -18,28 +24,153 @@
+        }
+    }
+
+    private void decreaseNumOfQuestions(int remainingNumOfQuestions) {
```
As above.

samsontmr commented on src/client/java/teammates/client/scripts/scalabilitytests/InstructorFeedbackResultsPageScalabilityTest.java:
```diff
@@ -18,28 +24,153 @@
+        Set<String> feedbackResponsesKeys = testDataMax.feedbackResponses.keySet();
+        String giver;
+        String recipient;
+        String questionId;
```
Is it necessary to declare these variables outside the foreach loop?

samsontmr commented on src/client/java/teammates/client/scripts/scalabilitytests/InstructorFeedbackResultsPageScalabilityTest.java:
```diff
@@ -18,28 +24,153 @@
+            if (studentsEmails.contains(giver)
+                    && studentsEmails.contains(recipient)
+                    && questionsNumbers.contains(questionId)) {
+                if (!testData.feedbackResponses.containsKey(key)) {
```
Consider combining these two if statements. You can abstract the condition out if it becomes over long.

Malkone commented:
@samsontmr Thank you! I made all proposed changes with following comments:
- Instead of renaming method ```decreaseNumOfStudents()```, I renamed method ```increaseNumOfStudents()``` to ```addStudents()``` because this name correspond more to what this method does. After this change can I keep name ```decreaseNumOfStudents()```? Because this name fits best for what this method does.
- Also in method ```updateFeedbackResponses()``` I removed temporary variables completely and now I pass results of get() method directly to if condition

Malkone commented:
@samsontmr Thank you!

Malkone commented:
@samsontmr I made one more correction - I remove excessive checking if ```testData.feedbackResponses.containsKey(key)```, because put() method will replace existing data anyway

samsontmr commented:
Strange, I remember commenting a few days ago. 

@whipermr5 this shouldn't result in extra costs, right?

whipermr5 commented on src/client/java/teammates/client/scripts/scalabilitytests/InstructorFeedbackResultsPageScalabilityTest.java:
```diff
@@ -18,28 +24,145 @@
+        // modify set of responses for each test according to its sets of students and questions
+        updateFeedbackResponses();
+
         removeAndRestoreDataBundle(testData);
```
This PR is neat; unfortunately it doesn't do anything to reduce the number of Datastore operations at all because this line still does the remove and restore operation on the entire `testData` bundle.

To correct, the approach has to be changed. Do a persist (without remove) data bundle for adding new data (this should be what the test measures). At the very end of the test, do a remove data bundle for all data. You'll need to add some methods in `BaseTestCaseWithBackDoorApiAccess` to support this.

EDIT: I just realised the total number of operations would still be the same, even with that approach. @samsontmr I'm not sure how you intended to reduce the number?

damithc replied:
If the test is measuring the scalability of read operations, we can persist data once and not touch them again unless the data needs to change. That means writing data should not be part of the test, but a separate script that is run manually whenever we need. 

For measuring scalability of write operations, we can need to separate the _primary_ objects whose write are being measured from _supplementary_ objects that simply need to be there for the write to succeed. e.g. The session object needs to be in the datastore for an operation of writing responses to succeed. After that, the supplementary object can be written once using a script, similar to the case above.

samsontmr replied:
In this case, it is a read operation scalability test. Instead of writing the entire dataset at the start of the test, perhaps we can check if the data required already exists in the datastore? The data isn't changed in the test per se, but in between tests, when the number of students/questions changes.

Malkone replied:
Please confirm if I got it right: before loading data from json file to ```DataBundle``` instance with necessary data for scalability test, we want to check if such instance already exists? 
Maybe we can use separate varianle instead of ```testData``` declared in superclass ```BaseUiTestCase```. So we will check if this variable is not empty and contains all necessary data. In this case we also don't need to generate json file if required ```DataBundle``` exists because this file will not be read in this case. 

> The data isn't changed in the test per se, but in between tests, when the number of students/questions changes.
Here i don't quite got it - so should the data be updated (added and removed) between test when number of students and questions change? Or should each test instead of modifying ```DataBundle``` should just read  the required number of students, questions and feedback responses from persisted data?

whipermr5 replied:
> Or should each test instead of modifying DataBundle should just read the required number of students, questions and feedback responses from persisted data?

I think it's this one, but will wait for @samsontmr to confirm :)

Malkone replied:
I'm confused about the following  - as google app engine documentation describes write operations are made when entities are created. In my previous refactoring I created such entities as Student, Question and QuestionResponse only once, but for each test I created many additional sets - one on adding and one on descreasing number of students, same on questions, and even 3 sets in mentioned method ```updateFeedbackResponses()``` before each test. These are all entities as I understood, so this might be the reason why the number of calls to database remains very high. It seems that I need combine the following goals in refactoring:
- check before uploading data from json file and not upload it if exists
- minimize creating additional sets when preparing data for each set

samsontmr replied:
> * check before uploading data from json file and not upload it if exists
* minimize creating additional sets when preparing data for each set

Yes I believe you are on the right track @malkone

whipermr5 replied:
@Malkone The actual entities are only created by the storage component when `removeAndRestoreDataBundle(testData)` is called. Before that, the objects you add to `testData` are only Attributes (data transfer objects) created in memory to indicate to the storage component what kind of entities to create.

Malkone replied:
@whipermr5 does it mean that my refactoring of code before calling this method  ```removeAndRestoreDataBundle(testData)``` will not influence the number of entities and writing call? In this case, seems like this storage component should not create entities each time in ```removeAndRestoreDataBundle(testData)``` method to minimize operations. This method should be upadted but I can't figure out how to check if entities were written to Datastore already and exist there. Could you please give a hint where to search for these operations?

whipermr5 replied:
> does it mean that my refactoring of code before calling this method removeAndRestoreDataBundle(testData) will not influence the number of entities and writing call?

Yes, exactly. You should only call `removeAndRestoreDataBundle` to add the data after determining that the data does not currently exist on the server. You can use `verifyPresentInDatastore` to check (catch the assertion error thrown if it is absent).

Malkone replied:
Friends, please help with one more question: there is method ```verifyPresentInDatastore(DataBundle data)``` and inside it calls ```verifyAbsentInDatastore(EntityAttributes entity)``` only for some of the inclusive objects from DataBundle: ```accounts```, ```instructors```, ```courses``` and ```students```. While in our scalability test we read and write DataBundle with more inclusive object, like ```feedbackQuestions``` and ```feedbackResponses```. So ```verifyPresentInDatastore(DataBundle data)``` check separately for ```account```, ```instructor```, ```course``` and ```student``` if it exists in Datastore, but do not checks their connection as parts of some single DataBundle. 
And in our test method ```removeAndRestoreDataBundle(testData)``` reads and writes one single complete set of data. In this case: 
- do we necessarily want to test reading whole DataBundle from Datastore? Then seems like I need to create another overloaded method ```verifyPresentInDatastore(DataBundle data)``` to check all the objects in my DataBundle if they persist on Datastore. And if somehow only one entity is absent, progra,, will be forced to read and write whole DataBundle (same current number of operatios will remain)
- or can I modify the test and make it read only ```Students```, ```FeedbackQuestions``` and ```FeedbackResponses``` (data, which is actually scaled). In this case I will use current methods from ```Backdoor``` class like ```getStudent()``` and ```getFeedbackQuestion()``` to read ```Students```, ```FeedbackQuestions``` and ```FeedbackResponses```, which persisit in Datastore. In this case rest of objects from testData will not be tested.
Which is preferrable?

Malkone commented:
@whipermr5 @samsontmr Friends, please help with one more question: there is method ```verifyPresentInDatastore(DataBundle data)``` and inside it calls ```verifyAbsentInDatastore(EntityAttributes entity)``` only for some of the inclusive objects from DataBundle: ```accounts```, ```instructors```, ```courses``` and ```students```. While in our scalability test we read and write DataBundle with more inclusive object, like ```feedbackQuestions``` and ```feedbackResponses```. So ```verifyPresentInDatastore(DataBundle data)``` check separately for ```account```, ```instructor```, ```course``` and ```student``` if it exists in Datastore, but do not checks their connection as parts of some single DataBundle. 
And in our test method ```removeAndRestoreDataBundle(testData)``` reads and writes one single complete set of data. In this case: 
- do we necessarily want to test reading whole DataBundle from Datastore? Then seems like I need to create another overloaded method ```verifyPresentInDatastore(DataBundle data)``` to check all the objects in my DataBundle if they persist on Datastore. And if somehow only one entity is absent, progra,, will be forced to read and write whole DataBundle (same current number of operatios will remain)
- or can I modify the test and make it read only ```Students```, ```FeedbackQuestions``` and ```FeedbackResponses``` (data, which is actually scaled). In this case I will use current methods from ```Backdoor``` class like ```getStudent()``` and ```getFeedbackQuestion()``` to read ```Students```, ```FeedbackQuestions``` and ```FeedbackResponses```, which persisit in Datastore. In this case rest of objects from testData will not be tested.
Which is preferrable?

whipermr5 commented:
@Malkone The second one (use the backdoor methods). As for number of calls, checking everything once at the start is still better than the current remove and refresh on every change.

Malkone commented:
@whipermr5 understand, thank you! I'll need to change also the ways how ```DataBundle``` for this test is generated, because ```BackDoor``` methods to check if a ```Student``` exists in Datastore checks for specific studentEmail, so we cannot any more generate random first and last name of students, we'll need to make these names constant for every test. I'll take some popular names for that

Malkone commented:
I seem to have done like discussed this thread, I replaced ```removeAndRestore()``` method because it writes whole ```DataBundle```. Instead I use newly created method ```verifyOrPersistTestDataToDatastore()``` which tries to get entity from datastore, and if fails - it adds this entity there. 

In result,  I get testing error, which says that I use illegal reference to ```teammates.test.driver.BackDoor```. I moved these methods to class ```BaseTestCaseWithDatastoreAccess``` but error remains. 

Then I tried to use logic like used for ```doRemoveAndRestoreDataBundle()``` - I announced these methods abstract in class ```BaseTestCaseWithDatastoreAccess``` and implemented them in class ```BaseTestCaseWithBackDoorApiAccess``` - but again not succesfull because I would need to implement methods in other subclasses. 

So how should I call ```createStudent()``` to persists entity to datastore if it's absent?

whipermr5 commented:
@Malkone looks like you managed to fix the error yourself; good job! Is this ready for review?

Malkone commented:
@whipermr5 thanks, I really solved that issue and everything seems to work now, but I'm not sure if the testing itself now is performed as expected. New logic works in the following way: 

- first, I create a test Databunle for each test as done previously, since we discussed that operations before removeAndRestore() method does not influence datastore
- second, I check each ```Student```, ```Question``` and ```FeedbackResponse``` from test Databundle, if they exist on Datastore,
- and while checking, if I find that entity does not exist - I write it to datastore with methods ```doPutStudent()```, ```doPutFeedbackQuestion()``` and ```doPutFeedbackResponse()```, which I created in class ```BaseTestCaseWithBackDoorApiAccess```.
So, at this stage I checked each entity and wrote those, which were absent in datastor. Obviously, now I don't need to use ```removeAndRestore()```, because it will make read and write calls for each entity, which we aim to decrease. 
So I made method ```readTestDataFromDatastore()``` but it almost totally repeats my ```verifyOrPersistTestDataToDatastore()```, just without writing entity if it is absent. Because ```getEntity()``` method in class ```BaseTestCaseWithDatastoreAccess``` is private and is used by those ```verify``` methods. 

Is it correct way to actually test scalability?

whipermr5 commented:
@Malkone nice work! `readTestDataFromDatastore()` isn't required at all; the real testing happens in `loginToInstructorFeedbackResultsPage`.

Malkone commented:
@whipermr5 Thanks a lot for your great support! Without your help I couldn't do this!
Now I removed ```readTestDataFromDatastore()```

whipermr5 commented on src/client/java/teammates/client/scripts/scalabilitytests/InstructorFeedbackResultsPageScalabilityTest.java:
```diff
@@ -18,20 +25,150 @@
+    private void verifyOrPersistTestDataToDatastore() {
+        for (StudentAttributes student : testData.students.values()) {
+            try {
+                verifyPresentInDatastore(student);
```
I'm getting a `NullPointerException` here; looks like `verifyPresentInDatastore` isn't converting it to an `AssertionError`; my bad. Could you change this to use `getStudent` instead?

whipermr5 replied:
Similarly for the feedback questions and responses.

whipermr5 replied:
Even after clearing those issues, the test fails at https://github.com/TEAMMATES/teammates/pull/7289/files#diff-b40ea85664f4cc1c1c4378f8415c1a1fR191 because `CFResultsScT.instr` doesn't exist in the Datastore.

Malkone replied:
Yes, I understood my mistake now - I should check every test data if it exists in datastore. So I corrected 3rd issue - added checks for not only for ```Instructor```, but also for ```Account``` and ```Course```.  As for feedbackSession - there is a method in ```BackDoor``` to check it in datastor, but there isn't a method ```createFeedbackSession()``` - I would try to create but not really sure how to write ```createFeedbackSession()``` correctly (it seems to be more complicated that creating ```Student``` or ```Question```).

Also, I cannot access to ```getEntity()``` method from here, because it is private in class ```BaseTestCaseWithDatastoreAccess``` - will it be OK to make it public? 
Or maybe solving that 3rd issue also solves 1st and 2nd issues from your comments?
Could you please check now? And help to write ```createFeedbackSession()``` in ```BackDoor```? 
Thanks!

whipermr5 replied:
Don't use `verifyPresentInDatastore` as it throws an ugly `NullPointerException` when the data doesn't exist. You can use `getCourse`, `getAccount`, `getInstructor`, `getStudent` etc. instead of `getEntity`.

whipermr5 replied:
As for `createFeedbackSession`, it should be very similar to the existing backdoor method `createFeedbackResponse`, except the middle line of code should look like
```
dataBundle.feedbackSessions.put("dummy-key", feedbackSession);
```
I haven't tested it out though.

Malkone replied:
Thanks a lot! I made all these changes now

whipermr5 commented:
@Malkone I tried running the test but it's failing [here](https://github.com/TEAMMATES/teammates/pull/7289/files#diff-b40ea85664f4cc1c1c4378f8415c1a1fR233) due to no results to expand:
<img width="1392" alt="screen shot 2017-06-13 at 9 13 07 pm" src="https://user-images.githubusercontent.com/6811428/27084104-40f3ebc4-507d-11e7-9b20-d853de64bbe1.png">
Looks like the data still isn't being inserted correctly? You can try resetting your local database to a clean state to test it out (delete `src/main/webapp/WEB-INF/appengine-generated/local_db.bin`).

Malkone commented:
Thanks for pointing this out, I seem to have some problems with local database - I cannot login to dev server as administrator, I get this error from server: ```Persistent class "Class teammates.storage.entity.Account does not seem to have been enhanced.``` like described in this isse [Dev Server's Problem #7210](https://github.com/TEAMMATES/teammates/issues/7210). I checked all settings as provided in answers but still cannnot fix this problem..

I also don't have folder ```src/main/webapp/WEB-INF/appengine-generated/```, my ```src/main/webapp/WEB-INF/``` looks like on pic 
<img width="500" alt="untitled 2" src="https://user-images.githubusercontent.com/24193624/27199147-a8194c26-521d-11e7-9e37-a3292f3b07d5.png">

whipermr5 commented:
Are you using IntelliJ?

Malkone commented:
I use Eclipse neon.3

whipermr5 commented:
Do your settings look like this?
<img src="https://raw.githubusercontent.com/TEAMMATES/teammates/3842f3a840c546c5f7e7434d5f6f20470b11bec4/docs/images/setupguide-5.png">

whipermr5 commented on src/client/java/teammates/client/scripts/scalabilitytests/InstructorFeedbackResultsPageDataGenerator.java:
```diff
@@ -17,6 +17,7 @@
  */
 @SuppressWarnings("PMD.UnusedPrivateField")//Inner classes and their fields are only used for JSON object generation.
```
Can take this opportunity to change this to `@SuppressWarnings("unused")` to prevent lots of warnings in Eclipse. Also, add spaces around the `//`.

Malkone replied:
Thanks, I'll make this change!

Malkone replied:
@whipermr5 I corrected this now, sorry I missed it in previous iterations

whipermr5 commented:
@Malkone Nope that's fine; the name doesn't matter (it's input by the user when adding the SDK). What about your ORM settings?
<img width="661" alt="screen shot 2017-06-19 at 9 38 43 pm" src="https://user-images.githubusercontent.com/6811428/27287957-e3a26738-5537-11e7-8dd2-a73ace23f822.png">

Malkone commented:
@whipermr5 Hi, sorry for this delay with answer, I seem to have a little different ORM settings with additinal ```*.java``` at the end. Should I remove this ending and keep only ```src/main/java/teammates/storage/entity/*.java``` ?

Also I have some errors because some imports fail like ```import com.google.api.services.gmail.*``` and ```import com.google.api.client.extensions.*```

whipermr5 commented:
@Malkone Your ORM settings are fine (it shows you followed the instructions exactly 👍). To fix the dependencies, do `./gradlew resetEclipseDeps`. Also, is your compiler compliance setting Java 1.7?

Malkone commented:
@whipermr5 Thanks! This helped and those errors disappeared!  👍 One more issue now left with build path - in properties there is a notice ```Build path entry is missing: com.google.appengine.eclipse.core.GAE_CONTAINER/appengine-java-sdk-1.9.27``` and after build it notifies about one build problem:  ```Unbound classpath container: 'App Engine SDK [missing]' in project 'teammates'```

whipermr5 commented:
@Malkone Is your setup working now? Do run your test and if it works without error, let me know :)

Malkone commented:
@whipermr5 Thanks a lot! Yes, everything works fine now after I upgraded app engine, I'm now making tests to investigate how to solve this pull request, will notify you, when make my changes will work :)

Malkone commented:
@whipermr5 hi, can you please assist - to test scalability I'm trying to login to the app on my dev server as ```CFResultsScT.instr@gmail.tmt``` but it fails, I see notice "You are not authorized to view this page. Logout and return to main page". How can I login as this Instructor so that I could check scalability test results for that session. 
For me this test is unusual because it is not like Travis CI test case - it is not in the list for Travis CI or Local test, so I don't understand when this test actually performs?

- when I build my source so I only need to enter this profile ```CFResultsScT.instr``` and check if Session results are OK
- or I need to login as ```CFResultsScT.instr``` and some kind of initiate this test from account profile?

whipermr5 commented:
@Malkone You need to manually run the test (run as TestNG test); it's a scalability test so it's not meant to be run in CI, only manually. The first time you run it, it'll create the accounts (your own code that adds instructors if not present).

whipermr5 commented:
@Malkone You need to make sure you call `verifyOrPersistTestDataToDatastore()` once at the start of the test to ensure all the data has been loaded into the Datastore.

Malkone commented:
@whipermr5 Thanks! Now testing works great and browser is logging that Instructor and doing everything else by itself)) So only jow I can fully test my changes to this pull request!
I'll give you a notice when my code works as supposed.

whipermr5 commented:
@Malkone Is the code ready? :)

Malkone commented:
@whipermr5 Hi! Yes, it seems to be working now - I left the test running during the night because it take more than 4 hours to complete the test. So in the morning I got successful result:

<img width="1279" alt="untitled" src="https://user-images.githubusercontent.com/24193624/28002917-00574686-6542-11e7-9679-b0fea690e80f.png">

Malkone commented:
@whipermr5 the only trouble I have with Firefox - after I ran the test once now when eclipse starts Firefox, on instructor page there are already present all 10 questions for all 20 students (though the test only started for 5 students and 1 question). And Firefox quits unexpectingly when I try to clear history. 

It is because I installed version 46.0.1 as separate instance and I set-up different profiles for different Firefox version, so that Firefox asks which profile to choose at lanch. And despite I cancelled choosing default profile, eclipse starts Firefox without choosing any profile. And it quits unexpectingly when I try to clear history in Firefox instance launched by eclipse.

Malkone commented:
Oops, travis testing failed with some PMD rule violation... I'll correct this in the evening

whipermr5 commented:
@Malkone That doesn't sound like Firefox history; it's a fact that the data is meant to be loaded into the datastore once during the very first run, then reused for all subsequent tests.

whipermr5 commented:
I believe Firefox doesn't store any history between runs. It is the test that should be modified; I think there needs to be a separate data set for each magnitude (e.g. one session that has 5 questions, another that has 10 etc) and all the sets should always be in the Datastore after the very first run - that's what the whole issue is about.

Malkone commented:
@whipermr5 Thanks, I though about this option - that these Questions already in Datastore so they appear, i just didn't believe in it :) because after some test I deleted all content in folder WEB-INF/appengine-generated (because I had too much errors when running the test), but still Firefox opened with all Questions already there. Now I believe App Engine stores that data even in some different place if it remains after I restart my system.

Malkone commented:
@whipermr5 it seems to be working good now :)

Malkone commented:
I made the change so that checking data in datastore is performed only once in the beginning, but still it doesn't find feedback responses in datastore and writes each of them (4000 in total). I made logging to each add data operation and each takes around 1 sec, so the test goes nearly 70 minutes. For some reason program does not find feedback responses on datastore and writes them there while it finds students and questions and does not make write operations on them.

whipermr5 commented:
@Malkone Good job diagnosing the issue! I'll take a look at why it isn't finding the responses.

whipermr5 commented on src/client/java/teammates/client/scripts/scalabilitytests/InstructorFeedbackResultsPageScalabilityTest.java:
```diff
@@ -18,20 +33,196 @@
         // the actual test data is refreshed before each test method
     }
 
     private void refreshTestData(int numStudents, int numQuestions) {
```
This whole method (with all its helper methods) should be removed, as no modifications to the datastore should be done after the very first check. It is `testContentWithIncreasingLoad` that must be updated to load different pages (each corresponding to a different data set that already exists), instead of refreshing test data and loading the same page.

whipermr5 replied:
I'd actually recommend closing this PR and starting afresh, as this one is getting too long (plus the approach has completely changed).

whipermr5 replied:
@Malkone Well done so far! Your persistence is commendable 👍💯

Malkone replied:
@whipermr5 thanks a lot for good words! It is much dependant to your great support and patience to responde to all my sometimes silly questions :)

whipermr5 commented on src/client/java/teammates/client/scripts/scalabilitytests/InstructorFeedbackResultsPageScalabilityTest.java:
```diff
@@ -18,20 +33,196 @@
+        int count = 1;
+        for (FeedbackResponseAttributes feedbackResponse : dataToVerify.feedbackResponses.values()) {
+
+            if (getFeedbackResponse(feedbackResponse) == null) {
```
The `getFeedbackResponse` method checks the `FeedbackResponseAttributes` object `feedbackResponse` for `feedbackResponse.feedbackQuestionId`, `feedbackResponse.giver` and `feedbackResponse.recipient`. Are those values correctly populated in the test data?

Malkone replied:
@whipermr5 I made new PR and testing there - but still have this issue and cannot solve it( 
the data seems to be populated correctly, because it includes ```feedbackResponse.feedbackQuestionId```, ```feedbackResponse.giver``` and ```feedbackResponse.recipient```, it looks like this in JSON before populating:

`{
  "accounts": {},
  "courses": {},
  "instructors": {},
  "students": {},
  "feedbackSessions": {},
  "feedbackQuestions": {},
  "feedbackResponses": {
    "dummy-key": {
      "feedbackSessionName": "Open Session",
      "courseId": "CFResultsScT.CS2104",
      "feedbackQuestionId": "0",
      "feedbackQuestionType": "TEXT",
      "giver": "CFResultsScT.Name2.Student2@gmail.tmt",
      "recipient": "CFResultsScT.Name2.Student2@gmail.tmt",
      "responseMetaData": {
        "value": "Response."
      },
      "giverSection": "Section A",
      "recipientSection": "Section A"
    }
  },
  "feedbackResponseComments": {},
  "profiles": {},
  "adminEmails": {}
}` 

So it has lines like ```"feedbackQuestionId": "0"```, ```"giver": "CFResultsScT.Name2.Student2@gmail.tmt"``` and ```"recipient": "CFResultsScT.Name2.Student2@gmail.tmt"```, but still in method ```getFeedbackResponse(String feedbackQuestionId, String giverEmail,String recipient)``` in class ```BackDoor``` method ```makePostRequest(params)``` returns ```null``` - this method uses ```sendRequest(paramString, conn)``` and ```readResponse(conn)``` which I cannot investigate deeper, because these methods just send data to specified URL and then gains response from it, and I cannot check what is happening there...

Malkone replied:
And more confusing - when making this new test on new git branch I use different studnet names like ```Name1.Student1``` and ```Name2.Student2``` instead of previous version ```Student.One``` and  ```Student.One``` (because in that case all names were equal).
But now I see in browser for each Question feedback responses from new students ```Name1.Student1``` and ```Name2.Student2```, and also from old students ```Student.One``` and  ```Student.One```

whipermr5 replied:
Oh I got the answer when working on #2258. The feedback question ID isn't actually 0; it's populated only when it's persisted to the datastore. The `feedbackQuestionId` field in test JSON corresponds to the question number instead (the actual ID is then injected into it before persisting). You'll need to first get the feedback question via session name, course ID and question number (these fields can all be found in the JSON) using the `getFeedbackQuestion` method, read the actual ID from it, then replace the value of the `feedbackQuestionId` field of the 'FeedbackResponseAttributes` object with the actual ID before passing it to `getFeedbackResponse`.

whipermr5 replied:
The old data still in the datastore is no surprise, as the data still exists in your `local_db.bin`, which isn't tracked by git.

whipermr5 replied:
As for debugging backdoor methods, check out `BackDoorServlet`. Output can be found in the dev server log (instead of the test log).

Malkone replied:
@whipermr5 thanks, it works! The real IDs for ```FeedbackQuestionAttributes ``` are not numbers at all and look like ```agt0ZWFtbWF0ZXN2NHIdCxIQRmVlZGJhY2tRdWVzdGlvbhiAgICAgICoCQw``` :) 
It is confusing that in JSON file for ```feedbackResponses``` the variable is named ```feedbackQuestionId``` but it shows ```questionNumber``` instead of ID:

> "response1166": {
      "giver": "CFResultsScT.Name17.Student17@gmail.tmt",
      "recipient": "CFResultsScT.Name9.Student9@gmail.tmt",
      "feedbackQuestionId": "2",
      "courseId": "CFResultsScT.CS2104",
      "feedbackQuestionType": "TEXT",
      "feedbackSessionName": "Open Session",
      "giverSection": "Section A",
      "recipientSection": "Section A",
      "responseMetaData": {
        "value": "Response."
      }

whipermr5 replied:
Yes; take a look at 'BackDoorLogic.persistDataBundle` inject methods to see why :)

Malkone replied:
Yes, I investigated it and ```injectRealIds()``` method, used inside it :) it is still strange to me, why then inner class ```Response``` has only field ```feedbackQuestionId``` to which question number is written, only to be changed to question ID when this response is written to datastore...  It would be logical to have separate fields ```feedbackQuestionId``` and ```feedbackQuestionNumber```. Maybe I just don't know all the details why it is needed

whipermr5 commented:
Closing since you've continued work in #7777. You can delete "second attempt" from that PR's title.
